client/src/main/java/io/pravega/client/stream/ReaderGroupConfig.java

```diff
@@ -119,12 +166,60 @@ public ReaderGroupConfig build() {
            checkArgument(startingStreamCuts != null && startingStreamCuts.size() > 0,
                    "Stream names that the reader group can read from cannot be empty");
 
-           startingStreamCuts.forEach((s, streamCut) -> {
+           //endStreamCuts is an optional configuration. Initialize if it is null.
+           if (endingStreamCuts == null) {
+               endingStreamCuts = Collections.emptyMap();
+           }
+
+           //validate start and end StreamCuts
+           validateStreamCut(startingStreamCuts);
+           validateStreamCut(endingStreamCuts);
+
+           //basic check to verify if endStreamCut > startStreamCut.
+           validateStartAndEndStreamCuts(startingStreamCuts, endingStreamCuts);
+
+           return new ReaderGroupConfig(groupRefreshTimeMillis, automaticCheckpointIntervalMillis,
+                   startingStreamCuts, endingStreamCuts);
+       }
+
+       private void validateStartAndEndStreamCuts(Map<Stream, StreamCut> startStreamCuts,
+                                                  Map<Stream, StreamCut> endStreamCuts) {
+           endStreamCuts.entrySet().stream().filter(e -> !e.getValue().equals(StreamCut.UNBOUNDED))
+                                .forEach(e -> {
+                               if (startStreamCuts.get(e.getKey()) != StreamCut.UNBOUNDED) {
+                                   verifyStartAndEndStreamCuts(startStreamCuts.get(e.getKey()), e.getValue());
+                               }
+                           });
+       }
+
+       /**
+        * Verify if startStreamCut comes before endStreamCut.
+        */
+       private void verifyStartAndEndStreamCuts(final StreamCut startStreamCut, final StreamCut endStreamCut) {
+           final Map<Segment, Long> startPositions = startStreamCut.asImpl().getPositions();
+           final Map<Segment, Long> endPositions = endStreamCut.asImpl().getPositions();
+           //check offsets for overlapping segments.
+           startPositions.keySet().stream().filter(endPositions::containsKey)
+                        .forEach(s -> Preconditions.checkArgument(startPositions.get(s) <= endPositions.get(s),
+                                "Segment offset in startStreamCut should be <= segment offset in endStreamCut."));
+
+           val fromSCSummary = startPositions.keySet()
+                                             .stream().collect(summarizingInt(Segment::getSegmentNumber));
+           val toSCSummary = endPositions.keySet()
+                                         .stream().collect(summarizingInt(Segment::getSegmentNumber));
+           //basic check to 
+           Preconditions.checkArgument(fromSCSummary.getMin() <= toSCSummary.getMin(),
```

Comment:
Isn't it sufficient to check that: `fromSCSummary.getMax() <= toSCSummary.getMin()`?

Reply:
```
  +-------+------+-------
     |       |   8  |
     |   2   +------|
     |       |   7  |   10
     +-------+ -----|
     |       |   6  |
     |  1    +------+-------
     |       |   5  |
     +-------+------|
     |       |   4  |   9
     |  0    +------|
     |       |   3  |
     +-------+------+--------
```
Valid `fromStreamCut(0, 1, 7, 8)` and `toStreamCut(3, 4, 5, 10)` will break the check `fromSCSummary.getMax() <= toSCSummary.getMin()`